drv/riscv-plic/src/main.rs

```diff
@@ -0,0 +1,190 @@
+    };
+}
+
+fn irq_assigned(irq: u32) -> bool {
+    return unsafe { HUBRIS_IRQ_TASK_LOOKUP.contains(InterruptNum(irq)) };
+}
+
+#[repr(C)]
+struct ServerImpl {}
+
+impl idl::InOrderRiscvIntCtrlImpl for ServerImpl {
+    /// Disables the selected interrupt on the PLIC.
+    fn disable_int(
+        &mut self,
+        msg: &userlib::RecvMessage,
+        irq: u32,
+    ) -> Result<(), RequestError<RiscvIntCtrlError>> {
+        let owner: InterruptOwner = InterruptOwner {
+            task: msg.sender.index() as u32,
+            notification: irq,
+        };
+
+        match get_task_irqs(owner) {
+            Ok(irqs) => {
+                let plic = unsafe { &mut *PLIC_REGISTER_BLOCK };
+                plic.mask(0, irqs[0].0.try_into().unwrap());
+
+                return Ok(());
+            }
+            Err(()) => return Err(Runtime(RiscvIntCtrlError::IRQUnassigned)),
+        }
+    }
+
+    /// Enables the selected interrupt on the PLIC.
+    fn enable_int(
+        &mut self,
+        msg: &userlib::RecvMessage,
+        irq: u32,
+    ) -> Result<(), RequestError<RiscvIntCtrlError>> {
+        let owner: InterruptOwner = InterruptOwner {
+            task: msg.sender.index() as u32,
+            notification: irq,
+        };
+
+        match get_task_irqs(owner) {
+            Ok(irqs) => {
+                let plic = unsafe { &mut *PLIC_REGISTER_BLOCK };
+                plic.unmask(0, irqs[0].0.try_into().unwrap());
+
+                return Ok(());
+            }
+            Err(()) => return Err(Runtime(RiscvIntCtrlError::IRQUnassigned)),
+        }
+    }
+
+    /// Completes the interrupt on the PLIC, allowing for a new one to come
+    /// through.
+    fn complete_int(
+        &mut self,
+        msg: &userlib::RecvMessage,
+        irq: u32,
+    ) -> Result<(), RequestError<RiscvIntCtrlError>> {
+        let owner: InterruptOwner = InterruptOwner {
+            task: msg.sender.index() as u32,
+            notification: irq,
+        };
+
+        match get_task_irqs(owner) {
+            Ok(irqs) => {
+                let plic = unsafe { &mut *PLIC_REGISTER_BLOCK };
+                plic.complete(0, irqs[0].0.try_into().unwrap());
+
+                return Ok(());
+            }
+            Err(()) => return Err(Runtime(RiscvIntCtrlError::IRQUnassigned)),
+        }
+    }
+}
+
+impl idol_runtime::NotificationHandler for ServerImpl {
+    // The PLIC is only interested in interrupt notifications from the kernel
+    fn current_notification_mask(&self) -> u32 {
+        return PLIC_IRQ;
+    }
+
+    // An interrupt has come in.
+    // NOTE: Currently, the driver assumes that all interrupts come in on
+    //       Context 0.
+    fn handle_notification(&mut self, _bits: u32) {
+        let plic = unsafe { &mut *PLIC_REGISTER_BLOCK };
+        loop {
+            let irq: u32 = match plic.claim(0) {
+                Some(irq) => core::primitive::u16::from(irq) as u32,
+                None => break,
```

Comment:
Here and in the rest of the checks below, there is a break to exit out of the loop. What if there are more interrupts to claim? Shouldn't this be a continue instead of break?

Reply:
The first break should be there as a claim returning 0 means there are no more interrupts. The latter two (lines 135 and 148) probably should be `continue` instead; those are cases where an unassigned interrupt has come in.

Reply:
Makes sense.